crates/iota-json-rpc-types/src/lib.rs

```diff
@@ -55,3 +57,175 @@ impl<T, C> Page<T, C> {
+            object_id,
+            version,
+            digest,
+        }
+    }
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase", tag = "bcsEncoding")]
+#[serde(from = "MaybeTaggedBcsName")]
+pub enum BcsName {
+    Base64 {
+        #[serde_as(as = "Base64")]
+        #[schemars(with = "Base64")]
+        #[serde(rename = "bcsName")]
+        bcs_name: Vec<u8>,
+    },
+    Base58 {
+        #[serde_as(as = "Base58")]
+        #[schemars(with = "Base58")]
+        #[serde(rename = "bcsName")]
+        bcs_name: Vec<u8>,
+    },
+}
+
+impl BcsName {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self::Base64 { bcs_name: bytes }
+    }
+
+    pub fn bytes(&self) -> &[u8] {
+        match self {
+            BcsName::Base64 { bcs_name } => bcs_name.as_ref(),
+            BcsName::Base58 { bcs_name } => bcs_name.as_ref(),
+        }
+    }
+
+    pub fn into_bytes(self) -> Vec<u8> {
+        match self {
+            BcsName::Base64 { bcs_name } => bcs_name,
+            BcsName::Base58 { bcs_name } => bcs_name,
+        }
+    }
+}
+
+#[allow(unused)]
+#[serde_as]
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", untagged)]
+enum MaybeTaggedBcsName {
+    Tagged(TaggedBcsName),
+    Base58 {
+        #[serde_as(as = "Base58")]
+        #[serde(rename = "bcsName")]
+        bcs_name: Vec<u8>,
+    },
+}
+
+#[serde_as]
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "bcsEncoding")]
+enum TaggedBcsName {
+    Base64 {
+        #[serde_as(as = "Base64")]
+        #[serde(rename = "bcsName")]
+        bcs_name: Vec<u8>,
+    },
+    Base58 {
+        #[serde_as(as = "Base58")]
+        #[serde(rename = "bcsName")]
+        bcs_name: Vec<u8>,
+    },
+}
+
+impl From<MaybeTaggedBcsName> for BcsName {
+    fn from(name: MaybeTaggedBcsName) -> BcsName {
+        let bcs_name = match name {
+            MaybeTaggedBcsName::Tagged(TaggedBcsName::Base58 { bcs_name })
+            | MaybeTaggedBcsName::Base58 { bcs_name } => bcs_name,
+            MaybeTaggedBcsName::Tagged(TaggedBcsName::Base64 { bcs_name }) => bcs_name,
+        };
+
+        // Bytes are already decoded, force into Base64 variant to avoid serializing to
+        // base58
+        Self::Base64 { bcs_name }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn bcs_name_test() {
+        let bytes = vec![0, 1, 2, 3, 4];
+        let untagged_base58 = r#"{"bcsName":"12VfUX"}"#;
+        let tagged_base58 = r#"{"bcsEncoding":"base58","bcsName":"12VfUX"}"#;
+        let tagged_base64 = r#"{"bcsEncoding":"base64","bcsName":"AAECAwQ="}"#;
+
+        println!(
+            "{}",
+            serde_json::to_string(&TaggedBcsName::Base64 {
+                bcs_name: bytes.clone()
+            })
+            .unwrap()
+        );
+
+        assert_eq!(
+            bytes,
+            serde_json::from_str::<BcsName>(untagged_base58)
```

Comment:
Do those from_str calls in this test all use the `MaybeTaggedBcsName` struct?
I'm trying to understand how it works.

Reply:
Yes. All the `from_str` calls first deserialize the JSON into a `MaybeTaggedBcsName` (as of the `#[serde(from = "MaybeTaggedBcsName")]` attribute). This intermediate struct uses attributes `#[serde_as(as = "Base64")]` or #[serde_as(as = "Base58")] to decode the `bcsName` field into a `Vec<u8>`. Then, in the `From` implementation, it simply extracts the already-decoded bytes and wraps them in the `Base64` variant.

This conversion via `MaybeTaggedBcsName` is only used during deserialization. When you serialize a `BcsName`, it uses always the Base64 variant.